network_analysis/trace-a-utility-network/src/main/java/com/esri/samples/trace_a_utility_network/TraceAUtilityNetworkController.java

```diff
@@ -0,0 +1,493 @@
+              if (identifiedFeature instanceof ArcGISFeature) {
+
+                // create element from the identified feature
+                UtilityElement utilityElement = utilityNetwork.createElement((ArcGISFeature) identifiedFeature);
+
+                // check if the network source is a junction or an edge
+                if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.JUNCTION) {
+
+                  // check if the feature has a terminal configuration and multiple terminals
+                  if (utilityElement.getAssetType().getTerminalConfiguration() != null) {
+                    UtilityTerminalConfiguration utilityTerminalConfiguration =
+                        utilityElement.getAssetType().getTerminalConfiguration();
+                    List<UtilityTerminal> terminals = utilityTerminalConfiguration.getTerminals();
+
+                    if (terminals.size() > 1) {
+                      // prompt the user to select a terminal for this feature
+                      Optional<UtilityTerminal> userSelectedTerminal = promptForTerminalSelection(terminals);
+
+                      // apply the selected terminal
+                      if (userSelectedTerminal.isPresent()) {
+                        UtilityTerminal terminal = userSelectedTerminal.get();
+                        utilityElement.setTerminal(terminal);
+                        // show the terminals name in the status label
+                        String terminalName = terminal.getName() != null ? terminal.getName() : "default";
+                        statusLabel.setText("Feature added at terminal: " + terminalName);
+
+                        // don't create the element if no terminal was selected
+                      } else {
+                        statusLabel.setText("No terminal selected - no feature added");
+                        return;
+                      }
+                    }
+                  }
+
+                } else if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.EDGE) {
+
+                  // get the geometry of the identified feature as a polyline, and remove the z component
+                  Polyline polyline = (Polyline) GeometryEngine.removeZ(identifiedFeature.getGeometry());
+
+                  // compute how far the clicked location is along the edge feature
+                  double fractionAlongEdge = GeometryEngine.fractionAlong(polyline, mapPoint, -1);
+                  if (Double.isNaN(fractionAlongEdge)) {
+                    new Alert(Alert.AlertType.ERROR, "Cannot add starting location / barrier here.");
+                    return;
+                  }
+
+                  // set the fraction along edge
+                  utilityElement.setFractionAlongEdge(fractionAlongEdge);
+
+                  // update the status label text
+                  statusLabel.setText("Fraction along edge: " + utilityElement.getFractionAlongEdge());
```

Comment:
Can we format the fraction so it stops at 3 or so decimal places?

Reply:
Done!